### PR TITLE
Fix MQTT impl to deliver pending messages when connected with MQTT CleanSession set to false

### PR DESCRIPTION
**Pull request by abhipsaMisra**
You can read about MQTT CleanSession behavior for C2D messages here: https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#receiving-cloud-to-device-messages

### CODE REVIEW

File: iothub/device/src/IDelegatingHandler.cs [hunk 1]
```diff
@@ -34,6 +34,10 @@ internal interface IDelegatingHandler : IContinuationProvider<IDelegatingHandler
 
         Task EnableReceiveMessageAsync(CancellationToken cancellationToken);
 
+        // This is to ensure that if device connects over MQTT with CleanSession flag set to false,
+        // then any message sent while the device was disconnected is delivered on the callback.
+        Task EnsurePendingMessagesAreDelivered(CancellationToken cancellationToken);
```
Comment (drwill-ms) [inline iothub/device/src/IDelegatingHandler.cs:39]: Async suffix?

File: iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs [hunk 2]
```diff
@@ -360,14 +367,16 @@ private Message ProcessMessage(Message message, bool hasMessage)
...
                     lock (_syncRoot)
                     {
-                        _messageQueue.TryDequeue(out message);
-                        message.LockToken = message.LockToken;
-                        if (_qos == QualityOfService.AtLeastOnce)
+                        if (_messageQueue.TryDequeue(out message))
                         {
-                            _completionQueue.Enqueue(message.LockToken);
-                        }
+                            message.LockToken = message.LockToken;
```
Comment (drwill-ms) [inline iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs:372]: What does this do?

File: iothub/device/src/IDelegatingHandler.cs (hunk 1 above)
Comment (abhipsaMisra) [reply iothub/device/src/IDelegatingHandler.cs:39]: we don't have any impl in here, just the signature

File: iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs [hunk 3]
```diff
@@ -360,14 +372,16 @@ private Message ProcessMessage(Message message, bool hasMessage)
...
+                            if (_qos == QualityOfService.AtLeastOnce)
+                            {
+                                _completionQueue.Enqueue(message.LockToken);
```
Comment (drwill-ms) [inline iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs:380]: What does enqueuing the lock token mean?

File: iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs (hunk 2 above)
Comment (abhipsaMisra) [reply iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs:377]: good point, this is unnecessary

File: iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs (hunk 3 above)
Comment (abhipsaMisra) [reply iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs:380]: this is used while marking a c2d message as complete. (MQTT supports only completing messages, while on AMQP we can reject and abandon as well)

Comment (drwill-ms) [reply iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs:380]: I mean, I generally think of enqueuing objects to be picked up by something on the other side. In this case, we're enqueuing the lock token, not an actually object (message).

Is this how we communicate back to the service that we complete a message, by enqueuing the lock token?

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**